proteus/MeshAdaptPUMI/MeshFields.cpp

```diff
@@ -0,0 +1,223 @@
+#include "MeshAdaptPUMI.h"
+#include <apfMesh.h>
+#include <apfShape.h>
+
+void MeshAdaptPUMIDrvr::freeField(apf::Field*& f)
+{
+  if (f) {
+    apf::destroyField(f);
+    f = 0;
+  }
+}
+
+void MeshAdaptPUMIDrvr::freeNumbering(apf::Numbering*& n)
+{
+  if (n) {
+    apf::destroyNumbering(n);
+    n = 0;
+  }
+}
+
+int MeshAdaptPUMIDrvr::transferFieldToPUMI(const char* name, double const* inArray,
+    int nVar, int nN)
+{
+  assert(nN == static_cast<int>(m->count(0)));
+  apf::Field* f = m->findField(name);
+  if (!strcmp(name, "coordinates")) {
+    assert(!f); /* coordinates are special, not found by regular findField */
+    f = m->getCoordinateField();
+    assert(f);
+  }
+  if (!f) {
+    assert(nVar == 1 || nVar == 3);
+    int valueType;
+    if (nVar == 1)
+      valueType = apf::SCALAR;
+    else
+      valueType = apf::VECTOR;
+    f = apf::createFieldOn(m, name, valueType);
+  }
+  apf::NewArray<double> tmp(nVar);
+  apf::MeshEntity* v;
+  apf::MeshIterator* it = m->begin(0);
+  while ((v = m->iterate(it))) {
+    int i = localNumber(v);
+    for(int j = 0; j < nVar; j++)
+      tmp[j] = inArray[i * nVar + j];
+/*
+if(!strcmp(name,"p"))    
+  tmp[0] =0 ; //pressure   
+else if(!strcmp(name,"velocity")){   
+  apf::Vector3 pt;
+  m->getPoint(v,0,pt);    
+  double Lz = 0.05;    
+  double Uinf = 2e-3;    
+  tmp[0] =0; //u   
+  tmp[1] = Uinf*pt[2]/Lz;    
+  tmp[2] =0;   
+ }    
+*/
+/*
+apf::Vector3 pt;
+m->getPoint(v,0,pt);    
+double Lz = 1;    
+double Ly = 5;   
+double Umax =1.5;
+if(!strcmp(name,"p"))
+  tmp[0] = 0.6*(1-pt[1]/Ly); //pressure starts at 1 and goes to 0    
+else if(!strcmp(name,"velocity")){   
+  tmp[0] = 0;    
+  tmp[1] = 4/Lz/Lz*Umax*(pt[2]*Lz-pt[2]*pt[2]);  //dpdy = 1/Ly   
+  tmp[2] = 0;
+}
+*/
+
+    apf::setComponents(f, v, 0, &tmp[0]); 
+  }
+  m->end(it);
+  return 0;
+}
+
+int MeshAdaptPUMIDrvr::transferFieldToProteus(const char* name, double* outArray,
+    int nVar, int nN)
+{
+  assert(nN == static_cast<int>(m->count(0)));
+  apf::Field* f = m->findField(name);
+  if (!f)
+    fprintf(stderr, "couldn't find field \"%s\"\n", name);
+  assert(f);
+  assert(apf::countComponents(f) == nVar);
+  apf::NewArray<double> tmp(nVar);
+  apf::MeshEntity* v;
+  apf::MeshIterator* it = m->begin(0);
+  while ((v = m->iterate(it))) {
+    int i = localNumber(v);
+    apf::getComponents(f, v, 0, &tmp[0]);
+    for(int j = 0; j < nVar; j++)
+      outArray[i * nVar + j] = tmp[j];
+  }
+  m->end(it);
+  return 0;
+}
+
+int MeshAdaptPUMIDrvr::transferPropertiesToPUMI(double* rho_p, double* nu_p, double *g_p)
+{ 
+ rho[0] = rho_p[0]; rho[1] = rho_p[1];
+ nu[0] = nu_p[0]; nu[1] = nu_p[1];
+ g[0] = g_p[0]; g[1] = g_p[1]; g[2] = g_p[2];
+ return 0;
+}
+
+int MeshAdaptPUMIDrvr::transferBCtagsToProteus(int* tagArray,int idx, int* ebN, int*eN_global,double* fluxBC)
+{
+/*
```

Comment:
This is a remnant of an attempt at passing the BCs specified by the user into the MeshAdapt portion of the code. It was functional for a serial case, but there were issues when running in parallel that were not resolved. I will comment the function out until it is usable.